### DISCUSSION TITLE
SNES Battletoads & Double Dragon (RareSNES) Samples / Playback / SF2

### DISCUSSION THREAD

**Discussion by ReaperMantis** [vgmtrans/vgmtrans#571] [Music formats]
Hello, I'm looking for some advice to point me in the right direction. I would like to export some SNES Battletoads & Double Dragon songs as MIDI with soundfonts. This game does not appear to be fully supported, however. VGMTrans is able to detect sequences and the RAREsnes format. It does not detect the Samples or Instrument Sets and is unable to play back the SPC files.

**Comment by ReaperMantis** [comment]
In the RareSnesScanner, the process is looking for a table that maps instrument numbers to indexes into the Source Directory. The scanner is using the following snippets to do this:
```
//; Donkey Kong Country SPC
// 0ac0: 4d        push  x
// 0ac1: f7 01     mov   a,($01)+y
// 0ac3: 5d        mov   x,a
// 0ac4: f5 e0 04  mov   a,$04e0+x         ; read SRCN table
// 0ac7: ce        pop   x
BytePattern RareSnesScanner::ptnReadSRCNTable(
	"\x4d\xf7\x01\x5d\xf5\xe0\x04\xce"
	,
	"xx?xx??x"
	,
	8);
```
and

```
...
  // Rare engine has a instrument # <--> SRCN # table, find it
  uint32_t ofsReadSRCNASM;
  if (!file->searchBytePattern(ptnReadSRCNTable, ofsReadSRCNASM)) {
    return;
  }
...
```
This BytePattern does not exist in Battletoads & Double Dragon, so the program immediately returns upon not finding this pattern. I traced the execution of Donkey Kong Country for a Program Change operation, and it did pass through the above code BytePattern and get to the table at 0x04e0.

I did the same for Battletoads & Double Dragon, and it does not resemble this at all. This is a call to perform a Program (and Volume Change). The program is changing the Kick sample to a Snare sample.

```
0C06  CE             POP X
0C07  8D 01          MOV Y,#$01
0C09  DB 3C          MOV $3C+X,Y        ;$01
0C0B  E8 00          MOV A,#$00
0C0D  D4 6C          MOV $6C+X,A        ;$00
0C0F  D4 2C          MOV $2C+X,A        ;$00
0C11  F7 04          MOV A,[$04]+Y      ;$0A This is the Program Change to instrument 0A (snare)
0C13  D5 34 02       MOV !$0234+X,A     ;$0B (kick) This is the Current Instrument that will be replaced by 0A
0C16  FC             INC Y
0C17  F7 04          MOV A,[$04]+Y      ;From here on, it's Volume Change information.
0C19  FC             INC Y
0C1A  60             CLRC
0C1B  97 04          ADC A,[$04]+Y
0C1D  5C             LSR A
0C1E  D5 44 02       MOV !$0244+X,A
0C21  D5 54 02       MOV !$0254+X,A
0C24  8F 04 04       MOV $04,#$04
0C27  5F ED 07       JMP !$07ED
```

Looking through the SPC file, I couldn't find a table that maps instrument numbers to SRCN at all. It seems to me that the track is indexing directly into the Source Directory.

Is it possible that this table doesn't exist and this game handles it differently than other RARE games?

**Comment by KungFuFurby** [comment]
Given that I've been doing tracing on these sound drivers across the entire commercial era, I'll explain what's going on here.

Every single game with a Rare sound driver build except for the two Donkey Kong Country sequels has a unique build (and sometimes there are version differences), and the VCMD set for each one is also different to some extent. I will give you some tracing notes going backwards from Donkey Kong Country.

For Battletoads-Double Dragon...
- The VxSRCN translation table does not yet exist. Instead, VCMD $01 directly sets a VxSRCN ID.
- VCMD $15 skips five parameter bytes. I determined this might have been an echo parameter command that was dummied out in this game. Donkey Kong Country had echo, but doesn't have one of the parameters: I suspect it's the EDL, since in Donkey Kong Country that's fixed to $04.
- VCMDs $2D-$30 don't exist at all.
- There is a byte missing from the song header. The byte in question is a SFX tempo variable.
- The sound driver doesn't even support stereo! Well, it does read two volume values, but it turns them into mono values.
- There is an ARAM map conflict: that is, variable ARAM overlaps the stack. Music playback may be affected. The most likely affected variable appears to be vibrato related.

**Comment by ReaperMantis** [comment]
Thanks for getting back to me! You have quite a history, and if anyone knows what's going on in these game, it's you!

This confirms my suspicions that there is no translation table and the $01 for 28 (Program/Vol change) sets the index to the instrument directly. I haven't delved deeper into the commands because the midi that gets generated seems --mostly-- correct by falling "forward" on the DKC event map, though it's worth a deeper look. I'm not going to think about the ARAM/stack overlap right now, I only started looking at SPC700 asm last week :P.

Regarding VGMtrans, in its current form, the RARE scanner is limited. The reliance on the VxSRCN translation table will prevent Battletoads-Double Dragon and Battlemaniacs from fully processing if they don't have one.

Does it makes sense to figure out what engine version is being used up front using something like [these sorting notes](https://raw.githubusercontent.com/loveemu/vgm-disasm/refs/heads/master/snes/Rare/Rare%20Build%20Sorting%20Notes.txt) or [this sorting list](https://sneslab.net/wiki/Rareware_SPC_Engine/Build_Sorting) from sneslab, or is that intentionally avoided to try to gain flexibility?  Currently, VGMtrans is using the routine for executing a VCMD for engine version identification, which on one hand is how Battletoads-Double Dragon can do anything at all (it shares the routine with DKC so that's how it gets processed), but on the other isn't totally correct for reasons you pointed out above.

**Comment by KungFuFurby** [comment]
Those happen to be my own sorting notes as well. I use these to quickly identify a build, and my usual method for doing so involves a pointer-based methodology. I do fall back to opcode-based comparisons sometimes as a result of inconsistent pointers or data in the middle of code.

I recommend doing something else from a code analysis standpoint. Also, the very earliest one on the list, Battletoads in Battlemaniacs (Beta and Final), has quite the set of differences.

...and actually, I have disassemblies of both of these. Check them out here (along with binaries of all of the remaining builds that weren't there before): https://github.com/KungFuFurby/vgm-disasm/tree/rare-update/snes/Rare

**Comment by ReaperMantis** [comment]
Wow, those are quite thoroughly commented too, thank you!

Can you elaborate on the recommendation to do something else?  For the most part, identifying the pointers works. Why not add more and tweak the flow?
